"""
ç®€å•çš„ gRPC æµ‹è¯•å®¢æˆ·ç«¯
ä½¿ç”¨ DEMO_MODE æµ‹è¯•æµå¼é€šä¿¡
"""
import asyncio
import grpc
import os
from loguru import logger

from app.gen.agent.v1 import agent_service_pb2, agent_service_pb2_grpc
from app.core.security import create_access_token
from app.config import settings


async def test_demo_mode():
    """
    æµ‹è¯• DEMO_MODE ä¸‹çš„æµå¼å¯¹è¯
    """
    logger.info("ğŸ§ª Testing gRPC StreamChat with DEMO_MODE...")

    async with grpc.aio.insecure_channel('localhost:50051') as channel:
        stub = agent_service_pb2_grpc.AgentServiceStub(channel)

        # ä½¿ç”¨é¢„è®¾çš„æ¼”ç¤ºå…³é”®è¯
        request = agent_service_pb2.ChatRequest(
            user_id="demo_user",
            session_id="demo_session",
            message="å¸®æˆ‘åˆ¶å®šé«˜æ•°å¤ä¹ è®¡åˆ’",  # è¿™æ˜¯ DEMO_MOCK_RESPONSES ä¸­çš„å…³é”®è¯
            user_profile=agent_service_pb2.UserProfile(
                nickname="æ¼”ç¤ºåŒå­¦",
                timezone="Asia/Shanghai",
                language="zh-CN"
            ),
            request_id="demo_req_001"
        )

        token = create_access_token({"sub": "demo_user"})
        
        # Use Internal API Key to bypass JWT secret mismatch issues in dev environment
        # This matches how Gateway calls Agent
        # Use settings.INTERNAL_API_KEY which is loaded from .env by Pydantic
        internal_key = settings.INTERNAL_API_KEY
        
        metadata = (
            ("authorization", f"Bearer {token}"),
            ("user-id", "demo_user"),
            ("x-trace-id", "demo_trace_001"),
            ("x-internal-api-key", internal_key), # Add internal key
        )

        try:
            logger.info(f"ğŸ“¤ Sending request: {request.message}")
            print("\n" + "=" * 70)
            print("ğŸ¤– AI Response:")
            print("=" * 70)

            response_count = 0
            full_text = ""

            async for response in stub.StreamChat(request, metadata=metadata):
                response_count += 1

                if response.HasField("delta"):
                    # æ‰“å°æµå¼æ–‡æœ¬
                    print(response.delta, end="", flush=True)
                    full_text += response.delta

                elif response.HasField("status_update"):
                    status = response.status_update
                    state_name = agent_service_pb2.AgentStatus.State.Name(status.state)
                    logger.info(f"\nğŸ“ [{state_name}] {status.details}")

                elif response.HasField("full_text"):
                    logger.info(f"\nâœ… Completed! Total length: {len(response.full_text)} chars")

                elif response.HasField("error"):
                    error = response.error
                    logger.error(f"\nâŒ Error: [{error.code}] {error.message}")
                    return False

            print("\n" + "=" * 70)
            logger.success(f"âœ… Test completed successfully!")
            logger.info(f"ğŸ“Š Statistics:")
            logger.info(f"   - Response chunks: {response_count}")
            logger.info(f"   - Total characters: {len(full_text)}")

            return True

        except grpc.RpcError as e:
            logger.error(f"âŒ gRPC error: {e.code()} - {e.details()}")
            return False
        except Exception as e:
            logger.error(f"âŒ Unexpected error: {e}", exc_info=True)
            return False


if __name__ == '__main__':
    success = asyncio.run(test_demo_mode())
    exit(0 if success else 1)
